python/pyspark/sql/functions.py

```diff
@@ -2049,132 +2051,13 @@ def map_values(col):
-
-    def _create_judf(self):
-        from pyspark.sql import SparkSession
-
-        spark = SparkSession.builder.getOrCreate()
-        sc = spark.sparkContext
-
-        wrapped_func = _wrap_function(sc, self.func, self.returnType)
-        jdt = spark._jsparkSession.parseDataType(self.returnType.json())
-        judf = sc._jvm.org.apache.spark.sql.execution.python.UserDefinedPythonFunction(
-            self._name, wrapped_func, jdt, self.pythonUdfType)
-        return judf
-
-    def __call__(self, *cols):
-        judf = self._judf
-        sc = SparkContext._active_spark_context
-        return Column(judf.apply(_to_seq(sc, cols, _to_java_column)))
-
-    def _wrapped(self):
-        """
-        Wrap this udf with a function and attach docstring from func
-        """
-
-        # It is possible for a callable instance without __name__ attribute or/and
-        # __module__ attribute to be wrapped here. For example, functools.partial. In this case,
-        # we should avoid wrapping the attributes from the wrapped function to the wrapper
-        # function. So, we take out these attribute names from the default names to set and
-        # then manually assign it after being wrapped.
-        assignments = tuple(
-            a for a in functools.WRAPPER_ASSIGNMENTS if a != '__name__' and a != '__module__')
-
-        @functools.wraps(self.func, assigned=assignments)
-        def wrapper(*args):
-            return self(*args)
-
-        wrapper.__name__ = self._name
-        wrapper.__module__ = (self.func.__module__ if hasattr(self.func, '__module__')
-                              else self.func.__class__.__module__)
-
-        wrapper.func = self.func
-        wrapper.returnType = self.returnType
-        wrapper.pythonUdfType = self.pythonUdfType
-
-        return wrapper
-
-
-def _create_udf(f, returnType, pythonUdfType):
-
-    def _udf(f, returnType=StringType(), pythonUdfType=pythonUdfType):
-        if pythonUdfType == PythonUdfType.PANDAS_UDF:
-            import inspect
-            argspec = inspect.getargspec(f)
-            if len(argspec.args) == 0 and argspec.varargs is None:
-                raise ValueError(
-                    "0-arg pandas_udfs are not supported. "
-                    "Instead, create a 1-arg pandas_udf and ignore the arg in your function."
-                )
-        udf_obj = UserDefinedFunction(f, returnType, pythonUdfType=pythonUdfType)
-        return udf_obj._wrapped()
-
-    # decorator @udf, @udf(), @udf(dataType()), or similar with @pandas_udf
-    if f is None or isinstance(f, (str, DataType)):
-        # If DataType has been passed as a positional argument
-        # for decorator use it as a returnType
-        return_type = f or returnType
-        return functools.partial(_udf, returnType=return_type, pythonUdfType=pythonUdfType)
-    else:
-        return _udf(f=f, returnType=returnType, pythonUdfType=pythonUdfType)
+class PandasUDFType(enum.Enum):
```

Comment:
I'm surprised this is not in python built-in library. What's the common way to implement enum in python? cc @HyukjinKwon

Reply:
This is in python built-in library after python3.4. It's not built-in for python2.x, but available through "enum34" package.

There are other ways to implement enum without enum34, but all of them are either complicated (using type()) or not as type-safe (using ints).

Reply:
Oh, I think we should definitely avoid adding a dependency .. I think that `setup.py` should only be applied to pip installation and I believe we need to have a manual dependency control in Python, for example, Py4J if I haven't missed any recent change or anything .. (see `./spark/python/lib`).

I plan to take a closer look related with it soon after we have what to consider as hard dependencies (for example, PyArrow, Pandas and etc).

> What's the common way to implement enum in python?

Up to my knowledge, just use class attributes is common.

> This is in python built-in library after python3.4. It's not built-in for python2.x, but available through "enum34" package.
>
> There are other ways to implement enum without enum34, but all of them are either complicated (using type()) or not as type-safe (using ints).

I think here type-safe is not a concern I believe .. do we really need enum here? sounds like a overkill even if it does not need the manual backport for Python 2.

Reply:
I can't currently take a closer look right now but please let me know @icexelloss if I missed your intention. Let me take a closer look tonight.

Reply:
@HyukjinKwon We don't have to use enum34 - using it makes the code more a little bit cleaner and more "modern" but it's not absolutely necessary. If we don't have the python2.x backport issue, I'd argue we should use enum34. However, if adding such dependency is not desirable it makes sense to just use class attribute for simplicity.

I can revert to just use int values as enum.

Reply:
I see. Yea, I would like to suggest to use class attributes for now. I don't want to add a dependency here .. We could separately discuss later if you feel strongly about this.